types/jest-when/index.d.ts

```diff
@@ -10,21 +10,41 @@
 
 /// <reference types="jest" />
 
-export interface WhenMock<T = any, Y extends any[] = any>
-  extends jest.MockInstance<T, Y> {
-  calledWith(...matchers: Y): this;
-  expectCalledWith(...matchers: Y): this;
-  mockReturnValue(value: T): this;
-  mockReturnValueOnce(value: T): this;
-  mockResolvedValue(value: jest.ResolvedValue<T>): this;
-  mockResolvedValueOnce(value: jest.ResolvedValue<T>): this;
-  mockRejectedValue(value: jest.RejectedValue<T>): this;
-  mockRejectedValueOnce(value: jest.RejectedValue<T>): this;
-  mockImplementation(fn: (...args: Y) => T): this;
-  mockImplementationOnce(fn?: (...args: Y) => T): this;
+export type ArgumentOrMatcher<ArgTypes extends any[]> = {
+    [Index in keyof ArgTypes]: ArgTypes[Index] | WhenMock<boolean, [ArgTypes[Index]]>;
+};
+
+export interface WhenMock<T = any, Y extends any[] = any> extends jest.MockInstance<T, Y> {
+    calledWith(allArgsMatcher: AllArgsMatcher<Y>): this;
+    calledWith(...matchers: ArgumentOrMatcher<Y>): this;
+    expectCalledWith(allArgsMatcher: AllArgsMatcher<Y>): this;
+    expectCalledWith(...matchers: ArgumentOrMatcher<Y>): this;
+    mockReturnValue(value: T): this;
+    mockReturnValueOnce(value: T): this;
+    mockResolvedValue(value: jest.ResolvedValue<T>): this;
+    mockResolvedValueOnce(value: jest.ResolvedValue<T>): this;
+    mockRejectedValue(value: jest.RejectedValue<T>): this;
+    mockRejectedValueOnce(value: jest.RejectedValue<T>): this;
+    mockImplementation(fn: (...args: Y) => T): this;
+    mockImplementationOnce(fn?: (...args: Y) => T): this;
+    defaultReturnValue(value: T): this;
+    defaultResolvedValue(value: jest.ResolvedValue<T>): this;
+    defaultRejectedValue(value: jest.RejectedValue<T>): this;
+    defaultImplementation(fn: (...args: Y) => T): this;
+}
+
+export interface AllArgsMatcher<Y> {
+    (args: Y, equals: jest.MatcherUtils['equals']): boolean;
+    // Internal, but needed to distinguish from normal callables
+    _isAllArgsFunctionMatcher: true;
+    _isFunctionMatcher: true;
 }
 
-export type When = <T, Y extends any[]>(fn: ((...args: Y) => T) | jest.MockInstance<T, Y>) => WhenMock<T, Y>;
+export interface When {
+    <T, Y extends any[]>(fn: ((...args: Y) => T) | jest.MockInstance<T, Y>): WhenMock<T, Y>;
+
+    allArgs<Y extends any[]>(matcher: (args: Y, equals: jest.MatcherUtils['equals']) => boolean): AllArgsMatcher<Y>;
```

Comment:
So this just makes a predicate into a branded predicate?

Reply:
Correct. It's necessary to distinguish between normal callables and branded callables in the `calledWith` and `expectCalledWith` overloads.

Reply:
Thanks for clarifying!